front/src/form/bsdd/components/appendix/Appendix2MultiSelect.tsx

```diff
@@ -0,0 +1,224 @@
+import { gql, useQuery } from "@apollo/client";
+import { InlineError } from "common/components/Error";
+import { formatDate } from "common/datetime";
+import useDebounce from "common/hooks/use-debounce";
+import { FieldArray, useFormikContext } from "formik";
+import { Form, Query, QueryAppendixFormsArgs } from "generated/graphql/types";
+import React, { useEffect, useMemo, useState } from "react";
+import formatWasteCodeEffect from "../waste-code/format-waste-code.effect";
+
+const APPENDIX2_FORMS = gql`
+  query AppendixForms($siret: String!, $wasteCode: String) {
+    appendixForms(siret: $siret, wasteCode: $wasteCode) {
+      id
+      readableId
+      emitter {
+        company {
+          name
+        }
+      }
+      wasteDetails {
+        code
+        name
+        quantity
+        packagingInfos {
+          type
+          other
+          quantity
+        }
+      }
+      signedAt
+      quantityReceived
+      processingOperationDone
+    }
+  }
+`;
+
+export default function Appendix2MultiSelect() {
+  const [wasteCodeFilter, setWasteCodeFilter] = useState("");
+  const debouncedWasteCodeFilter = useDebounce(wasteCodeFilter, 500);
+  const { values, setFieldValue, getFieldMeta } = useFormikContext<Form>();
+  const meta = getFieldMeta<Form[]>("appendix2Forms");
+  const [hasChanged, setHasChanged] = useState(false);
+
+  const { loading, error, data } = useQuery<
+    Pick<Query, "appendixForms">,
+    QueryAppendixFormsArgs
+  >(APPENDIX2_FORMS, {
+    variables: {
+      ...(debouncedWasteCodeFilter && debouncedWasteCodeFilter !== ""
+        ? { wasteCode: debouncedWasteCodeFilter }
+        : {}),
+      siret: values.emitter?.company?.siret ?? "",
+    },
+    skip: !values.emitter?.company?.siret,
+    fetchPolicy: "network-only",
+  });
+
+  // because { query { appendix2Forms } } does not return forms that
+  // have already been appended to a BSDD, we need to keep track of
+  // initial value of form.appendix2Forms when updating a BSDD
+  const appendix2Candidates = useMemo(() => {
+    const appendix2Forms = data?.appendixForms ?? [];
+    return [...(meta.initialValue ?? []), ...appendix2Forms].filter(f => {
+      return wasteCodeFilter?.length
+        ? f.wasteDetails?.code === wasteCodeFilter
```

Comment:
Je trouve dommage qu'il faille taper le code en entier pour avoir des résultats et que la liste soit vide entre le moment où je commence à taper et où je fini. Je suggère :

```suggestion
        ? f.wasteDetails?.code?.includes(wasteCodeFilter)
```

Reply:
Le filtre que tu mentionnes est là pour filtrer également sur les bordereaux initiaux annexés en cas d'update d'un bordereau de regroupement. L'autre partie du filtre s'effectue directement en modifiant le variable `wasteCode` sur la query `appendix2Forms` ce qui ne permet pas de gérer un match partiel, sauf à modifier le comportant du filtre. Je me dis qu'en fait on pourrait tout simplement filtrer uniquement côté client.